### PR TITLE
Store profiling in fit after reading it in once

### LINKED ISSUE #507: In-memory caching for profiler samples in CmdStanFit objects
**Is your feature request related to a problem? Please describe.**

Currently, CmdStanFit objects to not retain profiler samples in memory after the `$profiles()` method. That means there is not a way to serialize a CmdStanFit object with its profiler samples, which means the user needs to directly manage the CmdStan CSV files. 

**Describe the solution you'd like**

It would be great if the `$profiles()` method assigned the profiler samples somewhere in the CmdStanFit object so the whole object could be serialized without having to worry about the CSV files afterwards. 

**Describe alternatives you've considered**

I considered micromanaging the CSV files with the profiler samples, but this is a bit low level for [`stantargets`](https://github.com/ropensci/stantargets) specifically.

**Additional context**

The original issue and a reproducible example are located at https://github.com/ropensci/stantargets/discussions/41.

``` r
library(cmdstanr)
#> This is cmdstanr version 0.4.0.9000
#> - Online documentation and vignettes at mc-stan.org/cmdstanr
#> - CmdStan path set to: /Users/c240390/.cmdstanr/cmdstan-2.26.1
#> - Use set_cmdstan_path() to change the path
callr::r(function() {
  library(cmdstanr)
  mcmc_program <- write_stan_file(
    'data {
    int<lower=0> N;
    int<lower=0,upper=1> y[N];
  }
  parameters {
    real<lower=0,upper=1> theta;
  }
  model {
    profile("likelihood") {
      y ~ bernoulli(theta);
    }
  }
  generated quantities {
    int y_rep[N];
    profile("gq") {
      y_rep = bernoulli_rng(rep_vector(theta, N));
    }
  }
  ')
  mod_mcmc <- cmdstan_model(mcmc_program)
  data <- list(N = 10, y = c(1,1,0,0,0,1,0,1,0,0))
  fit <- mod_mcmc$sample(data = data, seed = 123, refresh = 0)
  profiles <- fit$profiles()
  draws <- fit$draws()
  saveRDS(fit, "fit.rds")
})
#> NULL

fit <- readRDS("fit.rds")
str(fit$draws())
#>  'draws_array' num [1:1000, 1:4, 1:12] -8.18 -8.2 -8.22 -8.28 -8.19 ...
#>  - attr(*, "dimnames")=List of 3
#>   ..$ iteration: chr [1:1000] "1" "2" "3" "4" ...
#>   ..$ chain    : chr [1:4] "1" "2" "3" "4"
#>   ..$ variable : chr [1:12] "lp__" "theta" "y_rep[1]" "y_rep[2]" ...

fit$profiles()
#> Error: No profile files found. The model that produced the fit did not use any profiling.
```

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Summary

Fixes #507 

@wlandau would you mind checking that this plays nice with stantargets. I am 95% sure it should but just in case. Also no rush, whenever you have time.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting 
(this will be you or your assignee, such as a university or company): 
Rok Češnovar

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: R/fit.R [hunk 1]
```diff
@@ -752,13 +754,15 @@ CmdStanFit$set("public", name = "return_codes", value = return_codes)
...
-  for (f in self$profile_files()) {
-    profiles[[i]] <- data.table::fread(f, data.table = FALSE)
-    i <- i + 1
-  }
-  profiles
+  if (is.null(private$profiles_)) {
+    private$profiles_ <- list()
+    i <- 1
+    for (f in self$profile_files()) {
+      private$profiles_[[i]] <- data.table::fread(f, integer64 = "character", data.table = FALSE)
```
Comment (rok-cesnovar) [inline R/fit.R:761]: The `integer64 = "character"` change is unrelated, but I recently did a complete reinstall of my R environment and profiles() threw this warning:

```
Warning messages:
1: In require_bit64_if_needed(ans) :
  Some columns are type 'integer64' but package bit64 is not installed. Those columns will print as strange looking floating point data. There is no need to reload the data. Simply install.packages('bit64') to obtain the integer64 print method and print the data again.
```

There is no real reason to have thread IDs as numeric/integer so this change seems more reasoanble then requiring bit64.

File: tests/testthat/test-profiling.R [hunk 2]
```diff
@@ -4,7 +4,6 @@ if (not_on_cran()) {
 
 test_that("profiling works if profiling data is present", {
   skip_on_cran()
-  skip_if(cmdstan_version() < "2.26.0")
```
Comment (rok-cesnovar) [inline tests/testthat/test-profiling.R:7]: I removed this skips as we always tests with the latest version and I think this is not needed actually.

**Comment by wlandau** [comment]
Thanks, @rok-cesnovar! I tried your PR, and it will definitely allow me to fix https://github.com/ropensci/stantargets/discussions/41.

**Comment by rok-cesnovar** [comment]
Great, thanks. In that case this is ready for review.

File: R/fit.R (hunk 1 above)
Comment (jgabry) [reply R/fit.R:761]: Sounds good

**Review by jgabry — APPROVED** [review]
Looks good, thanks! Will merge now.

**Pull request merged**